File: solutions/java/src/main/java/com/epi/SearchAPairSortedArray.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.*;

public class SearchAPairSortedArray {
  // @include
  private static class IndexPair {
    public Integer index1;
    public Integer index2;

    public IndexPair(Integer index1, Integer index2) {
      this.index1 = index1;
      this.index2 = index2;
    }
  }
  public static IndexPair findPairSumK(List<Integer> A, int k) {
    IndexPair result = findPosNegPair(A, k);
    if (result.index1 == -1 && result.index2 == -1) {
      return k >= 0
          ? findPairUsingComp(A, k,
                              new Comparator<Integer>() {
                                @Override
                                public int compare(Integer o1, Integer o2) {
                                  if (o1 < o2) {
                                    return 0;
                                  }
                                  return o1.compareTo(o2);
                                }
                              })
          : findPairUsingComp(A, k, new Comparator<Integer>() {
            @Override
            public int compare(Integer o1, Integer o2) {
              if (o1 >= o2) {
                return 0;
              }
              return o1.compareTo(o2);
            }
          });
    }
    return result;
  }

  private static IndexPair findPairUsingComp(List<Integer> A, int k,
                                             Comparator<Integer> comp) {
    IndexPair result = new IndexPair(0, A.size() - 1);
    while (result.index1 < result.index2 &&
           comp.compare(A.get(result.index1), 0) == 0) {
      result.index1 = result.index1 + 1;
    }
    while (result.index1 < result.index2 &&
           comp.compare(A.get(result.index2), 0) == 0) {
      result.index2 = result.index2 - 1;
    }

    while (result.index1 < result.index2) {
      if (A.get(result.index1) + A.get(result.index2) == k) {
        return result;
      } else if (comp.compare(A.get(result.index1) + A.get(result.index2), k) ==
                 0) {
        do {
          result.index1 = result.index1 + 1;
        } while (result.index1 < result.index2 &&
                 comp.compare(A.get(result.index1), 0) == 0);
      } else {
        do {
          result.index2 = result.index2 - 1;
        } while (result.index1 < result.index2 &&
                 comp.compare(A.get(result.index2), 0) == 0);
      }
    }
    return new IndexPair(-1, -1); // No answer.
  }

  private static IndexPair findPosNegPair(List<Integer> A, int k) {
    // result.first for positive, and result.second for negative.
    IndexPair result = new IndexPair(A.size() - 1, A.size() - 1);
    // Find the last positive or zero.
    while (result.index1 >= 0 && A.get(result.index1) < 0) {
      result.index1 = result.index1 - 1;
    }

    // Find the last negative.
    while (result.index2 >= 0 && A.get(result.index2) >= 0) {
      result.index2 = result.index2 - 1;
    }

    while (result.index1 >= 0 && result.index2 >= 0) {
      if (A.get(result.index1) + A.get(result.index2) == k) {
        return result;
      } else if (A.get(result.index1) + A.get(result.index2) > k) {
        do {
          result.index1 = result.index1 - 1;
        } while (result.index1 >= 0 && A.get(result.index1) < 0);
      } else { // A[result.first] + A[result.second] < k.
        do {
          result.index2 = result.index2 - 1;
        } while (result.index2 >= 0 && A.get(result.index2) >= 0);
      }
    }
    return new IndexPair(-1, -1); // No answer.
  }
  // @exclude

  public static void main(String[] args) {
    Random rand = new Random();
    for (int times = 0; times < 10000; ++times) {
      int n;
      if (args.length >= 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = rand.nextInt(10000) + 1;
      }
      List<Integer> A = new ArrayList<>();
      for (int i = 0; i < n; ++i) {
        A.add(rand.nextInt(19999) - 9999);
      }
      Collections.sort(A, new Comparator<Integer>() {
        @Override
        public int compare(Integer o1, Integer o2) {
          return Integer.valueOf(Math.abs(o1)).compareTo(Math.abs(o2));
        }
      });
      int k = rand.nextInt(19999) - 9999;
      // System.out.println(A);
      // System.out.println(k);
      IndexPair ans = findPairSumK(A, k);
      if (ans.index1 != -1 && ans.index2 != -1) {
        assert(A.get(ans.index1) + A.get(ans.index2) == k);
        System.out.println(A.get(ans.index1) + "+" + A.get(ans.index2) + "=" + k);
      } else {
        Collections.sort(A);
        int l = 0, r = A.size() - 1;
        boolean found = false;
        while (l < r) {
          if (A.get(l) + A.get(r) == k) {
            System.out.println(A.get(l) + "+" + A.get(r) + "=" + k);
            Collections.sort(A, new Comparator<Integer>() {
              @Override
              public int compare(Integer o1, Integer o2) {
                return Integer.valueOf(Math.abs(o1)).compareTo(Math.abs(o2));
              }
            });
            ans = findPairSumK(A, k);
            found = true;
            break;
          } else if (A.get(l) + A.get(r) < k) {
            ++l;
          } else {
            --r;
          }
        }
        System.out.println("no answer");
        assert(!found);
      }
    }
  }
}

File: static/solutions/Search_a_pair_sorted_array.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <functional>
#include <iostream>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::greater_equal;
using std::less;
using std::pair;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

struct IndexPair;
template <typename Compare>
IndexPair FindPairUsingCompare(const vector<int>&, int, Compare);
IndexPair FindPositiveNegativePair(const vector<int>&, int);

// @include
struct IndexPair {
  int index_1, index_2;
};

IndexPair FindPairSumK(const vector<int>& A, int k) {
  IndexPair result = FindPositiveNegativePair(A, k);
  if (result.index_1 == -1 && result.index_2 == -1) {
    return k >= 0 ? FindPairUsingCompare(A, k, less<int>())
                  : FindPairUsingCompare(A, k, greater_equal<int>());
  }
  return result;
}

template <typename Compare>
IndexPair FindPairUsingCompare(const vector<int>& A, int k, Compare comp) {
  IndexPair result = IndexPair{0, static_cast<int>(A.size() - 1)};
  while (result.index_1 < result.index_2 && comp(A[result.index_1], 0)) {
    ++result.index_1;
  }
  while (result.index_1 < result.index_2 && comp(A[result.index_2], 0)) {
    --result.index_2;
  }

  while (result.index_1 < result.index_2) {
    if (A[result.index_1] + A[result.index_2] == k) {
      return result;
    } else if (comp(A[result.index_1] + A[result.index_2], k)) {
      do {
        ++result.index_1;
      } while (result.index_1 < result.index_2 && comp(A[result.index_1], 0));
    } else {
      do {
        --result.index_2;
      } while (result.index_1 < result.index_2 && comp(A[result.index_2], 0));
    }
  }
  return {-1, -1};  // No answer.
}

IndexPair FindPositiveNegativePair(const vector<int>& A, int k) {
  // result.index_1 for positive, and result.index_2 for negative.
  IndexPair result = IndexPair{static_cast<int>(A.size() - 1),
                               static_cast<int>(A.size() - 1)};
  // Find the last positive or zero.
  while (result.index_1 >= 0 && A[result.index_1] < 0) {
    --result.index_1;
  }

  // Find the last negative.
  while (result.index_2 >= 0 && A[result.index_2] >= 0) {
    --result.index_2;
  }

  while (result.index_1 >= 0 && result.index_2 >= 0) {
    if (A[result.index_1] + A[result.index_2] == k) {
      return result;
    } else if (A[result.index_1] + A[result.index_2] > k) {
      do {
        --result.index_1;
      } while (result.index_1 >= 0 && A[result.index_1] < 0);
    } else {  // A[result.index_1] + A[result.index_2] < k.
      do {
        --result.index_2;
      } while (result.index_2 >= 0 && A[result.index_2] >= 0);
    }
  }
  return {-1, -1};  // No answer.
}
// @exclude

static void SimpleTest() {
  vector<int> A = {0, 0, -1, 2, -3, -3};
  IndexPair ans = FindPairSumK(A, 2);
  assert(ans.index_1 != -1);
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 10000; ++times) {
    int n;
    if (argc >= 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    uniform_int_distribution<int> dis(-10000, 10000);
    generate_n(back_inserter(A), n, [&] { return dis(gen); });
    sort(A.begin(), A.end(), [](int x, int y) { return abs(x) < abs(y); });
    int k = dis(gen);
    IndexPair ans = FindPairSumK(A, k);
    if (ans.index_1 != -1 && ans.index_2 != -1) {
      assert(A[ans.index_1] + A[ans.index_2] == k);
      cout << A[ans.index_1] << "+" << A[ans.index_2] << "=" << k << endl;
    } else {
      sort(A.begin(), A.end());
      int l = 0, r = A.size() - 1;
      bool found = false;
      while (l < r) {
        if (A[l] + A[r] == k) {
          cout << A[l] << "+" << A[r] << "=" << k << endl;
          found = true;
          break;
        } else if (A[l] + A[r] < k) {
          ++l;
        } else {
          --r;
        }
      }
      cout << "no answer" << endl;
      assert(!found);
    }
  }
  return 0;
}
